### ISSUE TITLE
Infinite loop in prepare_inner when a deref'd ref edit hits a locked leaf ref

### ISSUE THREAD

**Issue by afuno** [GitoxideLabs/gitoxide#2920]
`Repository::commit()` and `Repository::commit_as(.., "HEAD", ..)` never return — they spin at 100% CPU — when `refs/heads/<branch>.lock` is present. Passing the branch name directly instead of `HEAD` returns the expected `LockAcquire` error in about 100 ms, so the difference is entirely in the `deref: true` path.

## Cause

When a ref edit has `deref: true`, `RefEditsExt::extend_with_splits_of_symbolic_refs` splits it into two edits: `updates[0]` is `HEAD` with `parent_index: None`, and `updates[1]` is `refs/heads/main` with `parent_index: Some(0)`.

If the lock for the leaf ref cannot be acquired, `Transaction::prepare_inner` re-labels the error with the name from the parent chain. The loop that walks that chain never advances its cursor on the terminating branch:

```rust
// gix-ref/src/store/file/transaction/prepare.rs, lines 399-408 on main
let mut cursor = change.parent_index;
let mut ref_name = change.name();
while let Some(parent_idx) = cursor {
    let parent = &updates[parent_idx];
    if parent.parent_index.is_none() {
        ref_name = parent.name();   // cursor is not updated here
    } else {
        cursor = parent.parent_index;
    }
}
```

With `updates[1].parent_index == Some(0)` and `updates[0].parent_index == None`, the `is_none()` branch is taken on every iteration and `cursor` keeps its value forever. `change.name()` allocates, so the loop allocates and frees continuously — which is why this presents as a busy spin rather than a quiet deadlock.

## Reproduction

`Cargo.toml`:

```toml
[package]
name = "gix-deref-lock-repro"
version = "0.0.0"
edition = "2021"

[dependencies]
gix = { version = "=0.85.0", default-features = false, features = ["sha1", "tree-editor"] }
```

`src/main.rs`:

```rust
use std::time::{Duration, Instant};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Pass "HEAD" (deref: true) or "refs/heads/main" (deref: false).
    let target = std::env::args().nth(1).unwrap_or_else(|| "HEAD".into());

    let dir = std::env::temp_dir().join(format!("gix-repro-{}", target.replace('/', "_")));
    let _ = std::fs::remove_dir_all(&dir);
    std::fs::create_dir_all(&dir)?;

    let repo = gix::init_bare(&dir)?;
    std::fs::write(dir.join("HEAD"), "ref: refs/heads/main\n")?;

    let sig = gix::actor::Signature {
        name: "Repro".into(),
        email: "repro@example.com".into(),
        time: gix::date::Time::now_utc(),
    };
    let mut buf = gix::date::parse::TimeBuf::default();
    let sig_ref = sig.to_ref(&mut buf);
    let tree = gix::ObjectId::empty_tree(repo.object_hash());

    // An initial commit, so refs/heads/main exists and the next commit has a parent.
    let first = repo
        .commit_as(sig_ref, sig_ref, "HEAD", "first", tree, Vec::<gix::ObjectId>::new())?
        .detach();
    println!("initial commit: {first}");

    // The lock a crashed or SIGKILLed process leaves behind.
    let lock = dir.join("refs/heads/main.lock");
    std::fs::OpenOptions::new().write(true).create_new(true).open(&lock)?;
    println!("planted {}", lock.display());

    std::thread::spawn(|| {
        std::thread::sleep(Duration::from_secs(20));
        eprintln!("!! commit_as has not returned after 20s -- hung");
        std::process::exit(1);
    });

    println!("calling commit_as(.., {target:?}, ..) ...");
    let started = Instant::now();
    let result = repo.commit_as(sig_ref, sig_ref, target.as_str(), "second", tree, vec![first]);
    let elapsed = started.elapsed();

    match result {
        Ok(id) => println!("returned Ok({id}) in {elapsed:?}"),
        Err(e) => {
            println!("returned Err in {elapsed:?}");
            println!("  Display: {e}");
            println!("  Debug:   {e:?}");
        }
    }
    Ok(())
}
```

### Expected — `cargo run -- refs/heads/main`

```
initial commit: 84cf950c8f86f987be3407297bbc68adab860ec1
planted /tmp/gix-repro-refs_heads_main/refs/heads/main.lock
calling commit_as(.., "refs/heads/main", ..) ...
returned Err in 116.191667ms
  Display: A lock could not be obtained for reference "refs/heads/main"
  Debug:   ReferenceEdit(FileTransactionPrepare(LockAcquire { source: PermanentlyLocked {
             resource_path: "/tmp/gix-repro-refs_heads_main/refs/heads/main",
             mode: AfterDurationWithBackoff(100ms), attempts: 7 }, full_name: "refs/heads/main" }))
```

### Actual — `cargo run -- HEAD`

```
initial commit: e51c5a16408079531f29e67688e84141ebc6fd06
planted /tmp/gix-repro-HEAD/refs/heads/main.lock
calling commit_as(.., "HEAD", ..) ...
!! commit_as has not returned after 20s -- hung
```

Without the watchdog the process runs indefinitely. Sampling it shows the stack pinned in `edit_references_as` → `prepare_inner` → `Edit::name`, in a malloc/free cycle.

## Why this matters

A stale `.lock` file is not an exotic state. `gix-lock` has no staleness policy — it never inspects mtime or PID and never removes an existing lock — and `gix-tempfile`'s signal handlers are opt-in and cannot cover `SIGKILL` in any case. So one ungraceful termination can leave a repository where every subsequent `commit()` hangs forever instead of returning the actionable `PermanentlyLocked` error, whose message already tells the user the lockfile may need manual deletion.

For a long-running process the difference is significant: the failure mode is not a returned error the caller can surface, but a consumed CPU core, plus any application-level lock held around the commit being held for the remaining lifetime of the process.

## Versions

Reproduced with gix 0.85.0 / gix-ref 0.65.0 on macOS 26.5.2, rustc 1.93.1. The same code is present on `main` at the lines quoted above.

## Workaround

Pass the concrete ref name rather than `HEAD`, which avoids the symbolic split. Note this also changes the compare-and-swap constraint that `commit_as` derives from `MustExistAndMatch` to `ExistingMustMatch`, which may or may not matter depending on the caller.

## Possible fix

Advancing the cursor unconditionally terminates and still ends at the root of the parent chain:

```rust
let mut cursor = change.parent_index;
let mut ref_name = change.name();
while let Some(parent_idx) = cursor {
    let parent = &updates[parent_idx];
    ref_name = parent.name();
    cursor = parent.parent_index;
}
```

I am happy to open a PR with this plus a regression test if the approach looks right.

**Comment by Byron** [comment]
Thanks for reporting! Incredible this bug, in an error-handler of all places, survived this long!
